packages/flutter/test/widgets/editable_text_test.dart

```diff
@@ -1998,6 +1998,65 @@ void main() {
     expect(renderEditable.text.text, 'text composing text');
     expect(renderEditable.text.style.decoration, isNull);
   });
+
+  testWidgets('text selection handle visibility', (WidgetTester tester) async {
+    final GlobalKey<EditableTextState> editableTextKey =
+        GlobalKey<EditableTextState>();
+
+    const String testText = 'XXXXX          XXXXX';
+    final TextEditingController controller = TextEditingController(text: testText);
+
+    final Widget widget = MaterialApp(
+      home: Align(
+        alignment: Alignment.topLeft,
+        child: SizedBox(
+          width: 100,
+          child: EditableText(
+            key: editableTextKey,
+            controller: controller,
+            focusNode: FocusNode(),
+            style: Typography(platform: TargetPlatform.android).black.subhead,
+            cursorColor: Colors.blue,
+            backgroundCursorColor: Colors.grey,
+            selectionControls: materialTextSelectionControls,
+            keyboardType: TextInputType.text,
+          ),
+        ),
+      ),
+    );
+
+    await tester.pumpWidget(widget);
+
+    final EditableTextState state =
+        tester.state<EditableTextState>(find.byType(EditableText));
+
+    void verifyVisibility(bool leftVisible, bool rightVisible) {
+      final RenderEditable renderObject = state.selectionOverlay.renderObject;
+      expect(renderObject.selectionStartInView.value, equals(leftVisible));
+      expect(renderObject.selectionEndInView.value, equals(rightVisible));
+    }
+
+    // Select the first word. Both handles should be visible.
+    await tester.longPressAt(const Offset(20, 10));
+    await tester.pump();
+    verifyVisibility(true, true);
+
+    // Drag the text slightly so the first word is partially visible. Only the
+    // right handle should be visible.
+    await tester.drag(find.byKey(editableTextKey), const Offset(-20, 0));
+    await tester.pump();
+    verifyVisibility(false, true);
+
+    // Drag the text all the way to the left so the first word is not visible at
+    // all (and the second word is fully visible). Both handles should be
+    // invisible now.
+    await tester.drag(find.byKey(editableTextKey), const Offset(-400, 0));
+    await tester.pump();
+    verifyVisibility(false, false);
+
+    // TODO(sjindel): Test dragging in the other direction (blocked by
+    // https://github.com/flutter/flutter/issues/27948).
```

Comment:
please embrace the yak shave and fix that bug too, so that you can write the test

Reply:
Done.